packages/snowpack/src/logger.ts

```diff
@@ -1,35 +1,88 @@
 import * as colors from 'kleur/colors';
-import pino, {LoggerOptions} from 'pino';
-
-const LEVEL = {
-  TRACE: 10,
-  DEBUG: 20,
-  INFO: 30,
-  WARN: 40,
-  ERROR: 50,
-  FATAL: 60,
-};
+import {LoggerLevel, LoggerEvent, LoggerOptions} from './types/snowpack';
 
-/** http://getpino.io/#/docs/pretty */
-export function prettifier(options: LoggerOptions) {
-  return (inputData: pino.LogDescriptor) => {
-    // if the log starts with brackets, use that (e.g. [@snowpack/plugin-babel]); otherwise, use [snowpack]
-    const name = options.name ? `[${options.name}]` : '[snowpack]';
-    let msg = `${colors.dim(name)} ${inputData.msg}\n`; // add newline at end
+const levels: Record<LoggerLevel, number> = {
+  debug: 20,
+  info: 30,
+  warn: 40,
+  error: 50,
+  silent: 90,
+};
 
-    if (inputData.level === LEVEL.ERROR || inputData.level === LEVEL.FATAL) msg = colors.red(msg);
-    if (inputData.level === LEVEL.WARN) msg = colors.yellow(msg);
+/** Custom logger heavily-inspired by https://github.com/pinojs/pino with extra features like log retentian */
+class SnowpackLogger {
+  /** set the log level (can be changed after init) */
+  public level: LoggerLevel = 'info';
 
-    return msg;
+  private history: string[] = []; // this is immutable; must be accessed with Logger.getHistory()
+  private callbacks: Record<LoggerEvent, (message: string) => void> = {
+    debug: () => {},
+    info: () => {},
+    warn: () => {},
+    error: () => {},
   };
-}
 
-/** use Pino, a tiny logging library */
-export default function createLogger(opts: pino.LoggerOptions = {}) {
-  return pino({
-    ...opts,
-    name: opts.name || 'snowpack',
-    prettyPrint: {suppressFlushSyncWarning: true},
-    prettifier,
-  });
+  private log({level, name, message}: {level: LoggerEvent; name: string; message: string}) {
+    // test if this level is enabled or not
+    if (levels[this.level] > levels[level]) {
+      return; // do nothing
+    }
+
+    // format
+    let text = message;
+    if (level === 'warn') text = colors.yellow(text);
+    if (level === 'error') text = colors.red(text);
+    const log = `${colors.dim(`[${name}]`)} ${text}`;
+
+    // add to log history
+    this.history = this.history.concat(log);
+
+    // log
+    let logFn = console.log;
+    if (level === 'warn') logFn = console.warn;
+    if (level === 'error') logFn = console.error;
+    logFn(log);
+
+    // fire callback, if any
```

Comment:
❓ Here, we write to `console.*` and **ALSO** fire a callback. What if, instead, the callback overwrote the default behavior?

In `build` and `install` it doesn’t matter, but in dev this is what happens:

1. `logger.info()` is called.
2. as a result, `console.log()` is fired
3. the dev server erases the terminal
4. the dev server reprints the log history with `stdout.write()`.

Basically, in the dev server, should we just skip step #2?

Reply:
+1, There's probably a real risk of flash of content on a slower device, especially as the history array grows (console.log prints, pause, then gets cleared).

I think its fair that if you have something listening to it, it's do present it in a better way to the user, in which case don't log it yourself. 

If that's confusing, then we could also look at renaming `on` to something with less baggage (so that people don't confuse it with eventemitter). Maybe `logger.capture()` or even `.pipe()`

Reply:
Updated. Now `logger.on(` replaces the default logging behavior, preventing this flashing.